packages/dds/tree/src/simple-tree/arrayNode.ts

```diff
@@ -673,11 +515,35 @@ function createArrayNodeProxy(
 	return proxy;
 }
 
-abstract class CustomArrayNodeBase<const T extends ImplicitAllowedTypes> extends TreeNodeValid<
-	Iterable<InsertableTreeNodeFromImplicitAllowedTypes<T>>
-> {
+type Insertable<T extends ImplicitAllowedTypes> = readonly (
+	| InsertableTreeNodeFromImplicitAllowedTypes<T>
+	| IterableTreeArrayContent<InsertableTreeNodeFromImplicitAllowedTypes<T>>
+)[];
+
+abstract class CustomArrayNodeBase<const T extends ImplicitAllowedTypes>
+	extends TreeNodeWithArrayFeatures<T>
+	implements TreeArrayNode<T>
+{
+	// Indexing must be provided by subclass.
+	[k: number]: TreeNodeFromImplicitAllowedTypes<T>;
+
 	public static readonly kind = NodeKind.Array;
 
+	protected abstract get simpleSchema(): T;
+
+	#cursorFromFieldData(value: Insertable<T>): ITreeCursorSynchronous {
+		const content = contextualizeInsertedArrayContent(
+			value as readonly (InsertableContent | IterableTreeArrayContent<InsertableContent>)[],
+			getSequenceField(this),
+		);
+
+		// TODO: this is not valid since this is a value field schema, not a sequence one (which does not exist in the simple tree layer),
+		// but it works since cursorFromFieldData special cases arrays.
+		const simpleFieldSchema = normalizeFieldSchema(this.simpleSchema);
```

Comment:
this is sketchy, but is how it was implemented before this change (except I deduplicated the three copies of this logic into here and added this todo)